### PR TITLE
feat: cp comp auto register

### PR DESCRIPTION
**Pull request by sfwn**
#### What this PR does / why we need it:

feat: cp comp auto register

#### Which issue(s) this PR fixes:

[Related Erda Issue](https://erda.cloud/erda/dop/projects/387/issues/all?id=272285&issueFilter__urlQuery=eyJ0aXRsZSI6IuWKqOaAgeazqOWGjCIsInN0YXRlcyI6WzQ0MDYsNDQwN119&issueTable__urlQuery=eyJwYWdlTm8iOjEsICJwYWdlU2l6ZSI6MTB9&iterationID=772&type=TASK)

#### Specified Reivewers:

/assign @Effet

### CODE REVIEW

File: providers/component-protocol/cpregister/base/provider_name.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+		return "", "", "", fmt.Errorf("not standard provider key: %s", providerKey)
+	}
+	vv := strings.SplitN(ss[3], "@", 2)
+	if len(vv) == 2 {
+		compName = vv[0]
+		instanceName = vv[1]
+	} else {
+		compName = ss[3]
+		instanceName = compName
+	}
```
Comment (Effet) [inline providers/component-protocol/cpregister/base/provider_name.go:58]: should encapsulate a func like: splitCompAndInstance

File: providers/component-protocol/cpregister/base/provider_name.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+		ss := strings.SplitN(providerKey, ".", 3)
+		return "", ss[2], ss[2], nil
+	}
+	if !strings.HasPrefix(providerKey, componentProviderNamePrefix) {
+		return "", "", "", fmt.Errorf("invalid prefix")
+	}
+	ss := strings.SplitN(providerKey, ".", 4)
+	if len(ss) != 4 {
+		return "", "", "", fmt.Errorf("not standard provider key: %s", providerKey)
+	}
```
Comment (Effet) [inline providers/component-protocol/cpregister/base/provider_name.go:50]: should encapuslate a func like: remove prefix or what?

File: providers/component-protocol/cpregister/base/comp_register.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+		creator = func() servicehub.Provider { return &DefaultProvider{} }
+	}
+	// register to servicehub
+	servicehub.Register(providerName, &servicehub.Spec{Creator: creator})
+	// add to explicit provider creator map for hubListener to auto register to hub.config
+	cpregister.AllExplicitProviderCreatorMap[providerName] = creator
+
+	// generate creators compatible for IComponent and old CompRender
+	creators := func() creators {
+		switch creator().(type) {
```
Comment (Effet) [inline providers/component-protocol/cpregister/base/comp_register.go:60]: can we assign create() to a local variable?

**Review by Effet — CHANGES_REQUESTED** [review]

File: providers/component-protocol/cpregister/base/provider_name.go (hunk 1 above)
Comment (sfwn) [reply providers/component-protocol/cpregister/base/provider_name.go:57]: adjusted.

File: providers/component-protocol/cpregister/base/provider_name.go (hunk 2 above)
Comment (sfwn) [reply providers/component-protocol/cpregister/base/provider_name.go:49]: adjusted

File: providers/component-protocol/cpregister/base/comp_register.go (hunk 3 above)
Comment (sfwn) [reply providers/component-protocol/cpregister/base/comp_register.go:60]: it is a local variable, generated by anonymous function's invoking.

**Review by Effet — APPROVED** [review]

**Pull request merged**